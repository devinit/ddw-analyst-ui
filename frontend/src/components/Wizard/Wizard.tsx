import React, { FunctionComponent, Children, isValidElement, ReactNode } from 'react';
import { Card } from 'react-bootstrap';
import styled from 'styled-components';
import { WizardBody, WizardHeader, WizardNavigation, WizardNavigationItem } from '.';

interface WizardProps {
  nav?: WizardNav[];
}

export interface WizardNav {
  key: string;
  caption?: string;
  active?: boolean;
}

const StyledCard = styled(Card)`
  opacity: 1;
  box-shadow: 0 1px 4px 0 rgba(0, 0, 0, 0.14);
`;

const Wizard: FunctionComponent<WizardProps> = ({ children, nav }) => {
  const renderWizardNavigation = (): ReactNode => {
    if (nav && nav?.length) {
      const col = Math.ceil(12 / nav.length);

      return nav.map(({ active, caption, key }) => (
        <WizardNavigationItem
          key={key}
          active={active}
          eventKey={`#${key}`}
          className={`col-${col}`}
        >
          {caption}
        </WizardNavigationItem>
      ));
    }

    return null;
  };

  return (
    <div className="wizard-container">
      <StyledCard className="card-wizard" data-color="red">
        {Children.map(children, (child) =>
          isValidElement(child) && child.type === WizardHeader ? child : null,
        )}
        <WizardNavigation>{renderWizardNavigation()}</WizardNavigation>
        {Children.map(children, (child) =>
          isValidElement(child) && child.type === WizardBody ? child : null,
        )}
      </StyledCard>
    </div>
  );
};

export { Wizard };
