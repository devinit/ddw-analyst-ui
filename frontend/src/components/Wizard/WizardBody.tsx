import React, { FunctionComponent } from 'react';
import { Card, Tab } from 'react-bootstrap';

const WizardBody: FunctionComponent = ({ children }) => {
  return (
    <Card.Body>
      <Tab.Content>{children}</Tab.Content>
    </Card.Body>
  );
};

export { WizardBody };
