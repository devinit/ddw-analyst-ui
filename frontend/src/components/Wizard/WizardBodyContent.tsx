import React, { FunctionComponent, useState, useEffect } from 'react';
import { Tab, TabPaneProps } from 'react-bootstrap';

interface ComponentProps extends TabPaneProps {
  children?: React.ReactNode;
}
const WizardBodyContent: FunctionComponent<ComponentProps> = ({ children, ...props }) => {
  const [viewed, setViewed] = useState(props.active);
  useEffect(() => {
    if (props.active) {
      setViewed(true);
    }
  }, [props.active]);

  return (
    <Tab.Pane data-testid="wizard-tab-pane" {...props}>
      {viewed ? children : null}
    </Tab.Pane>
  );
};

export { WizardBodyContent };
