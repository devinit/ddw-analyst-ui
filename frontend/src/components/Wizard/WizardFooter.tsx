import React, { FunctionComponent } from 'react';
import { Card } from 'react-bootstrap';

type ComponentProps = {
  children?: React.ReactNode;
};
const WizardFooter: FunctionComponent<ComponentProps> = ({ children }) => {
  return <Card.Footer>{children}</Card.Footer>;
};

export { WizardFooter };
