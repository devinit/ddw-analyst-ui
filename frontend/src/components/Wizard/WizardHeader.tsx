import React, { FC } from 'react';
import { Card } from 'react-bootstrap';

type ComponentProps = {
  children?: React.ReactNode;
};
const WizardHeader: FC<ComponentProps> = ({ children }) => {
  return <Card.Header className="text-center">{children}</Card.Header>;
};

export { WizardHeader };
