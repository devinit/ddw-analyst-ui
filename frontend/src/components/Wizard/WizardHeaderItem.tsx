import React, { FunctionComponent } from 'react';
import { Nav, NavItemProps } from 'react-bootstrap';

const WizardHeaderItem: FunctionComponent<NavItemProps> = ({ children, ...props }) => {
  return <Nav.Item {...props}>{children}</Nav.Item>;
};

export { WizardHeaderItem };
