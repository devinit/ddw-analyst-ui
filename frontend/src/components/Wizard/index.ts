export * from './Wizard';
export * from './WizardHeader';
export * from './WizardNavigation';
export * from './WizardNavigationtem';
export * from './WizardBody';
export * from './WizardBodyContent';
export * from './WizardFooter';
