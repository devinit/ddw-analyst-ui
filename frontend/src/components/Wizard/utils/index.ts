import { WizardStep } from '..';

export const showPreviousButton = (steps?: WizardStep[], activeStep?: WizardStep): boolean => {
  if (steps && steps.length > 1 && activeStep) {
    return steps.findIndex((step) => activeStep?.key === step.key) > 0;
  }

  return false;
};

export const showNextButton = (steps?: WizardStep[], activeStep?: WizardStep): boolean => {
  if (steps && steps.length > 1 && activeStep) {
    return steps.findIndex((step) => activeStep.key === step.key) < steps.length - 1;
  }

  return false;
};

export const showFinishButton = (steps?: WizardStep[], activeStep?: WizardStep): boolean => {
  if (steps && steps.length > 1 && activeStep) {
    return steps.findIndex((step) => activeStep.key === step.key) === steps.length - 1;
  }

  return false;
};
