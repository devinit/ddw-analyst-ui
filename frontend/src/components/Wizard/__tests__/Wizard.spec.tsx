/**
 * @jest-environment jsdom
 */
import 'jest-styled-components';
import '@testing-library/jest-dom/extend-expect';
import React from 'react';
import * as TestRenderer from 'react-test-renderer';
import { Wizard } from '../Wizard';

describe('Wizard', () => {
  test('renders the default structure correctly', () => {
    const renderer = TestRenderer.create(<Wizard />).toJSON();

    expect(renderer).toMatchSnapshot();
  });

  test('renders with children', () => {
    const renderer = TestRenderer.create(
      <Wizard>
        <div>My Child</div>
      </Wizard>,
    ).toJSON();

    expect(renderer).toMatchSnapshot();
  });
});
