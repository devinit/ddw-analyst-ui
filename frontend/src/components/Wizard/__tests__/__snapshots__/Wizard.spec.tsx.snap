// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Wizard renders the default structure correctly 1`] = `
.c1.nav {
  border-radius: inherit !important;
  margin: inherit !important;
  padding: inherit !important;
}

.c0 {
  opacity: 1;
  box-shadow: 0 1px 4px 0 rgba(0,0,0,0.14);
}

<div
  className="wizard-container"
>
  <div
    className="c0 card-wizard card"
    data-color="red"
  >
    <div
      className="wizard-navigation"
    >
      <div
        className="c1 nav nav-pills"
        onKeyDown={[Function]}
        role="tablist"
      />
    </div>
    <div
      className="card-footer"
    >
      <div
        className="mr-auto"
      >
        <button
          className="btn-previous btn-fill btn-wd btn-default d-none btn btn-primary"
          disabled={false}
          onClick={[Function]}
          type="button"
        >
          Previous
        </button>
      </div>
      <div
        className="ml-auto"
      >
        <button
          className="btn btn-next btn-fill btn-wd d-none btn btn-danger"
          disabled={false}
          onClick={[Function]}
          type="button"
        >
          Next
        </button>
        <button
          className="btn btn-fill btn-wd d-none btn btn-danger"
          disabled={false}
          name="finish"
          onClick={[Function]}
          type="button"
        >
          Finish
        </button>
      </div>
      <div
        className="clearfix"
      />
    </div>
  </div>
</div>
`;

exports[`Wizard renders with children 1`] = `
.c1.nav {
  border-radius: inherit !important;
  margin: inherit !important;
  padding: inherit !important;
}

.c0 {
  opacity: 1;
  box-shadow: 0 1px 4px 0 rgba(0,0,0,0.14);
}

<div
  className="wizard-container"
>
  <div
    className="c0 card-wizard card"
    data-color="red"
  >
    <div
      className="wizard-navigation"
    >
      <div
        className="c1 nav nav-pills"
        onKeyDown={[Function]}
        role="tablist"
      />
    </div>
    <div
      className="card-footer"
    >
      <div
        className="mr-auto"
      >
        <button
          className="btn-previous btn-fill btn-wd btn-default d-none btn btn-primary"
          disabled={false}
          onClick={[Function]}
          type="button"
        >
          Previous
        </button>
      </div>
      <div
        className="ml-auto"
      >
        <button
          className="btn btn-next btn-fill btn-wd d-none btn btn-danger"
          disabled={false}
          onClick={[Function]}
          type="button"
        >
          Next
        </button>
        <button
          className="btn btn-fill btn-wd d-none btn btn-danger"
          disabled={false}
          name="finish"
          onClick={[Function]}
          type="button"
        >
          Finish
        </button>
      </div>
      <div
        className="clearfix"
      />
    </div>
  </div>
</div>
`;
